File: rest_server.rb
# ruby rest_server.rb -s Puma
require "sinatra"
require "mysql2"
require "msgpack"
require "oj"
require "dalli"

configure :production do
  set :server, :puma
end

$count = 0

$client = Mysql2::Client.new(:host => "localhost", :username => "root", :database => "test")
def fetch_mysql
  $count += 1
  results = $client.query("SELECT name,mail FROM user WHERE id=#{$count}")
  row = results.first
  return row
end

get '/' do
  content_type 'application/x-msgpack'
  MessagePack.pack(fetch_mysql())
end

get '/json' do
  content_type 'application/json'
  Oj.dump(fetch_mysql())
end

$dalli = Dalli::Client.new(["localhost:11211"])
def fetch_memcached
  $count += 1
  row = {}
  (row["name"],row["mail"]) = $dalli.get($count).split('|',2)
  return row
end

get '/mem' do
  content_type 'application/x-msgpack'
  MessagePack.pack(fetch_memcached())
end

get '/mem.json' do
  content_type 'application/json'
  Oj.dump(fetch_memcached())
end

File: rest_server.pl
# plackup -s Starman -E production rest_server.pl
use Plack::Builder;
use DBI;
use Data::MessagePack;
use JSON::XS;
use Cache::Memcached::Fast;

my $mp = Data::MessagePack->new();
my $dbh = DBI->connect('DBI:mysql:test:localhost:3307', 'root', '');
my $memd = new Cache::Memcached::Fast({
    servers => [ { address => 'localhost:11211' } ],
});
my $id = 0;

sub fetch_mysql {
    $id++;
    $id = 0 if($id > 10000);
    my $sth = $dbh->prepare("select name,mail from user where id = ?");
    $sth->execute($id);
    my $row = $sth->fetchrow_hashref;
    $sth->finish;
    return $row;
}

sub fetch_memcached {
    $id++;
    @dat = split(/\|/, $memd->get($id), 2);
    return {name => $dat[0], mail => $dat[1]};
}

builder {
    # enable 'AxsLog', combined => 1, logger => sub {};
    # enable 'AccessLog', logger => sub { };
    sub{
        my $env = shift;
        #use Data::Dumper;
        #warn Data::Dumper::Dumper $env;

        if($env->{REQUEST_URI} eq '/'){
            return [200,['Content-Type','application/x-msgpack'],[$mp->pack(fetch_mysql())]];
        } elsif($env->{REQUEST_URI} eq '/json'){
            return [200,['Content-Type','application/json'],[encode_json(fetch_mysql())]];
        } elsif($env->{REQUEST_URI} eq '/mem'){
            return [200,['Content-Type','application/x-msgpack'],[$mp->pack(fetch_memcached())]];
        } else {
            return [200,['Content-Type','application/json'],[encode_json(fetch_memcached())]];
        }
    }
};

